crates/autopilot/src/domain/settlement/transaction.rs

```diff
@@ -0,0 +1,73 @@
+//! Domain types for on-chain transactions that settle settlements.
+
+use {
+    super::Settlement,
+    crate::{
+        domain::{auction, eth},
+        infra,
+    },
+};
+
+/// A transaction that settles a settlement. Interacts with the settlement
+/// contract `settle` function.
+pub struct Tx {
+    settlement: Settlement,
+    transaction: Transaction,
+    #[allow(dead_code)]
+    receipt: Receipt,
+}
+
+impl Tx {
+    pub async fn new(tx: eth::TxId, eth: infra::Ethereum) -> Result<Self, Error> {
+        let (transaction, receipt) =
+            tokio::try_join!(eth.transaction(tx), eth.transaction_receipt(tx),)?;
+
+        let domain_separator = eth.contracts().settlement_domain_separator();
+        let settlement = Settlement::new(&transaction.input.0.clone().into(), domain_separator)?;
+        Ok(Self {
+            settlement,
+            transaction,
+            receipt,
+        })
+    }
+
+    pub fn auction_id(&self) -> auction::Id {
+        self.settlement.auction_id()
+    }
+
+    pub fn solver(&self) -> eth::Address {
+        self.transaction.solver
+    }
+}
+
+/// An on-chain transaction that settles a settlement.
+#[derive(Debug)]
+pub struct Transaction {
+    /// The hash of the transaction.
+    pub hash: eth::TxId,
+    /// The address of the solver that submitted the transaction.
+    pub solver: eth::Address,
+    /// The call data of the transaction.
+    pub input: eth::Calldata,
```

Comment:
Why do we still need the input if this type is exclusively used with the already parsed `Settlement` struct. This again creates some redundancy which we want to avoid (we want to craft domain models which accurately and minimally describe the information we need)

Reply:
Conceptually, you are right that, in this case, `eth::Calldata` is read once and forwarded to decoder and never used anymore.
But how would you then fetch the transaction object and forward the calldata immediately to decoder (without violating DDD principles)? We need to have a domain representation of transaction (`Transaction` object). 

Imo, there is some price we have to pay since we want to incrementally, through small PRs, introduce new domain objects and reuse them in old code. For example, calldata is also reused [here](https://github.com/cowprotocol/services/blob/main/crates/autopilot/src/on_settlement_event_updater.rs#L257). One could argue that I could just implement new persistence function to get transaction/receipt and keep the old functions around for this line to work, but then again, I would be left with duplicated functions for transaction fetching. Bottom line, I don't think we can nail the pixel perfect domain design before we completely finish the DDD refactoring in the autopilot.

Reply:
I'm ok defining a single transaction object (would not split between transaction and receipt) which is used together with the auction to create the solution object (and the two things combined may be called settlement)